[WIP] net: implement a StratumV2 Template Provider in core

Fi3 opened the pull request:
This PR provides a simple example of how to import and use the `sv2-ffi` Rust crate which exports the Stratum V2 (Sv2) functions required by a Template Provider as a C library.

## Stratum V2
[An introduction to Sv2](https://docs.google.com/document/d/1efN1JnAGiTw96LtfP70BngtP49SLtbMahPYSROfiMqA/edit?usp=sharing)

[Sv2 specifiaction](https://docs.google.com/document/d/1FadCWj-57dvhxsnFM_7X806qyvhR0u3i85607bGHxvg/edit)

## Template Provider
Sv2 defines a service (role) called the Template Provider (TP), whose functionality is defined as follows:
```
Generates the custom block templates to be passed to the Job Negotiator for eventual mining. 
This is usually a Bitcoin Core full node (or possibly another node implementation).
```

## Phase 1: PR Goals
This phase encompasses the concepts that must be agreed upon before proceeding with Phase 2.

1. Need agreement that implementing a TP in Bitcoin Core is a good idea.
2. Need agreement that the implementation of a TP in Bitcoin Core should use the Rust crates in [this workspace](https://github.com/stratum-mining/stratum).

## Phase 2: PR Implementation
This phase will commence once the concepts in Phase 1 are agreed upon. Phase 2 is the final goal of this project and encompasses the the implementation steps required for a functional TP in Bitcoin Core.

1. Extensive review of the `sv2-ffi` API and other Rust sources.
2. Review of the guix build process.
3. Find a location for the Sv2 crates (possibly under the [Bitcoin GitHub Account](https://github.com/bitcoin/)) to live _*1_.
4. Implement the TP.

## If TP in Core gets a concept NACK
A TP can be implemented as an independent service and can communicate with core via RPC.

## If Rust in Core gets a concept NACK
A TP can be implemented in core without using any Rust dependency.

_*1_ Right now the Rust sources are in this PR and are packaged by a script before doing the guix build. Ideally these sources should live in another project repository and should be packaged by the guix script which builds a downloadable binary.

UnidenifiedUser committed:
Add sv2 via guix

Fi3 commented:
To build it:
```
./contrib/guix/guix-build
env HOSTS='x86_64-linux-gnu' ./contrib/guix/guix-build
```

To test it:
```
 ./guix-build-bb059e4be395/distsrc-bb059e4be395-x86_64-linux-gnu/src/bitcoind
```

An example of how to import `sv2-ffi` in a generic c++ projecy can be found [here](https://github.com/stratum-mining/stratum/tree/main/examples/interop-cpp)

TheBlueMatt commented on src/rusty/build-for-guix.sh:
```diff
@@ -0,0 +1,36 @@
+mkdir ${SCRIPT_DIR}/target
+mkdir ${SCRIPT_DIR}/target/package
+
+cargo package --manifest-path ${SCRIPT_DIR}/sv2-ffi/Cargo.toml --allow-dirty
```
I don't think Bitcoin Core build should ever call cargo for rust stuff - we should call rustc directly with the relevant libraries.

Fi3 replied:
I'm gonna remove the cargo dependency. And subtree all the rust sources.

luke-jr commented:
Concept NACK for Rust, until a secure bootstrap without trusted third party binaries is practical.

Fi3 commented:
> Concept NACK for Rust, until a secure bootstrap without trusted third party binaries is practical.

Isn't that already achieved with guix?

UnidenifiedUser committed:
Remove cargo dependency

UnidenifiedUser committed:
Remove unneccesary sources

rrybarczyk committed:
Remove local TP rust crates to be replaced by a git subtree

rrybarczyk committed:
Add sv2 Rust crates for TP implementation

rrybarczyk committed:
Update guix builds to reflect src/rusty directory structure

ariard commented:
I think this work is interesting, that said it might need more thought on how to interface with Core.

AFAICT, the proposed PR just link the library in the build system and create a new template encoder in `bitcoind`. Though it doesn't introduce a new thread where the logic would co-run with other Core's current threads (`ThreadSocketHandler`, `ThreadOpenConnection`, ...). I initially followed this approach when I proposed AltNet with #18988. After gathering [feedbacks](https://gnusha.org/bitcoin-core-dev/2020-05-21.log) from other contributors, a separated process sounds a more flexible and safer approach.

Recently, I've made some progress in this direction by building on top of the [multiprocess](https://github.com/bitcoin/bitcoin/pull/10102) project (see the hacky branch https://github.com/ariard/bitcoin/commits/2021-07-altnet-lightning). Process separation is promising, though going further I'm aiming to have AltNet pluggin modules written in Rust, while interfacing in Core over the  Cap'n Proto newer interfaces. The current bottleneck is re-writing `libmultiprocess` in Rust, see https://github.com/chaincodelabs/libmultiprocess/issues/56 for more information. It's more work than expected though I've a branch half-written doing that, hope to make it public soon(tm).

I think such modules or the one proposed in the current PR could live beyond the Core repository to avoid burdening the maintenance effort and the concerns about Rust chain of trust, while consuming the stable interfaces to ease deployment across the ecosystem.

Glad to see progress on the miner infrastructure, that's really needed!

TheBlueMatt commented:
Strong Concept ACK. At a high level, this is incredibly important work for a few reasons:
 * having a good block template producer in Bitcoin Core is critical for moving Bitcoin's mining decentralization forward over the coming years, the importance of which cannot be understated,
 * Bitcoin Core having more visibility/control into how and when block templates get produced gives Bitcoin Core a ton more flexibility going forward to do things like pre-construct next-next-block templates in advance and push them out quicker, gives Bitcoin Core a ton more control of when transactions get included in a block by enabling "push" support for template updates instead of only being polled, etc,
 * the inefficiencies of getblocktemplate have been a real source of pain for pool operators for some time, I'm excited its going away.

As for multi-process, I don't think this is a good candidate for multi-process - block template building is incredibly latency-sensitive, plus needs full access to the mempool and block template construction. At most, I'd think we could think of doing the block template construction in Core and then handing the template off to a new process which does the StratumV2 protocol work, but (a) that adds a bunch more latency for no reason, and (b) you'd probably just use the StratumV2 protocol to communicate with that other binary, so why? In general, one reason to consider rust for the protocol side of this is precisely because you *don't* need to worry as much about it being in the same binary - the actual network logic itself isn't going to corrupt memory.

sipa commented:
Some thoughts.

* I don't object in general to adding a Rust-based *dependency*, as long as we can treat it as such. I don't personally know Rust, and suspect that many contributors/reviewers aren't as comfortable with Rust, so I think my criterion for this is whether it's something that we can treat as external, existing, known-to-work, code that's maintained separately - like say libevent or LevelDB, and not something that would expect detailed review from this project. Evidence that this same code is used in other contexts/projects would certainly help. The fact that this would be code that's only accessible through a local, optional, interface makes the safety analysis a lot easier.
* I haven't followed the state of affairs with StratumV2. If it's expected that a significant part of the mining ecosystem will end up depending on this protocol, I agree it makes sense to include it in Bitcoin Core, and agree with @TheBlueMatt that direct integration is the lowest-latency solution.
* I believe Guix indeed solves the concern of having bootstrapping the Rust compiler, at least to the same extent that our C++ compilers for release builds are bootstrapped too. Right, @dongcarl?

So conditionally on getting some confirmation on my assumptions above, (somewhat hesitant) Concept ACK.

luke-jr commented:
No, Guix does not solve the bootstrapping problem, because Guix itself has a bootstrapping problem. You can't get Guix going without running *their* trusted blobs.

sipa commented:
> No, Guix does not solve the bootstrapping problem, because Guix itself has a bootstrapping problem. You can't get Guix going without running their trusted blobs.

That's correct (for now), but the situation is not Rust-specific. Our release builds are created with Guix, so the C++ compiler used is subject to the same concerns.

dongcarl commented:
> * I believe Guix indeed solves the concern of having bootstrapping the Rust compiler, at least to the same extent that our C++ compilers for release builds are bootstrapped too. Right, @dongcarl?

Yes I think this is _exactly_ right. Building Rust in Guix is not any less bootstrappable than building anything else (C++) in Guix.

-----

Some context: Guix gained the ability to compile Rust v1.19 from mrustc v0.8 in 2018 ([blog post](https://guix.gnu.org/en/blog/2018/bootstrapping-rust/)), and this bootstrap method has been actively maintained, and now compiles Rust v1.29 from mrustc v0.9 (newer versions of rust (say v1.n) simply build using the previous version (v1.(n-1)) until it reaches v1.29).

michaelfolkson commented:
Concept ACK on implementing a Stratum v2 Template Provider in Core ignoring language. 

On the C++/Rust question it is dependent on other contributors being comfortable with it. This was discussed in today's Core dev [meeting](https://gnusha.org/bitcoin-core-dev/2021-09-23.log) and has come up repeatedly in the past e.g. https://github.com/bitcoin/bitcoin/issues/17090. I would guess a C++ PR would be an easier/quicker merge but there may be a Rust approach that can work.

luke-jr commented:
>That's correct (for now), but the situation is not Rust-specific. Our release builds are created with Guix, so the C++ compiler used for releases is subject to the same concerns.

We support users building their own, without using Guix's trust-required toolchain.

ariard commented:
> As for multi-process, I don't think this is a good candidate for multi-process - block template building is incredibly latency-sensitive, plus needs full access to the mempool and block template construction.

Ah I see the low-latency requirement, yes effectively IPC/serialization to access the mempool is likely going to be a worrying performance penalty.

That said, one advantage splitting nicely the block template/mempool subsystems behind interfaces like it has been done for GUI/wallet/node is the ability offered to the end-user to have either one fully-fledged binary or a collection of process in function what make sense as a deployment. 

One motivation could be to have fallback mempools connected to your block template construction in case of a malicious partition targeting your main one.

UnidenifiedUser committed:
Update build system

It conditionally compile the Template Provider logic into core if:
* the system that compile core has rustc at least at version 0.51
* configure is called `enable-template-provider`

Template Provider is optional, with default to no, so no rustc is required to compile core.
When compiled with guix the Template Provider is compiled in. As guix do support rustc.

Fi3 commented:
> If Rust is really a critical part of doing this feature, maybe we should look at how to offer the raw data necessary (in a performant way) to construct any number of block template formats outside of Core, by supplying data in a generic format over e.g. Unix sockets.

Hi @jamesob:
* why rust: main reason to use rust it to not rewrite an already working set of libraries, the libraries have been written to be used  both in core (for the implementation of the TP) and in rust (for the implementation of the other Sv2 roles) 
* how will Rust work with existing locks: do not know I will check
* cost of maintenance/additional complexity in the build system: IMHO is pretty insignificant.
  Support TP in the build system add:
    1. few lines in the config file in order to check if there is the flag `--enable-template-provider` and if the system that will build core has at least rustc 1.51.
    2. An m4 function pulled from the gnu repo ax_compare_version.
    3. A make file that call rustc with the correct flag and build the rust library then add the builded library to LDFLAGS so that is linked with bitocoind (this is the most complex addition to the build system and are only 159 LOC that do a very simple task)
    4. It modify 2 line in the guix build and add rustc 1.51 in the manifest
* how much third-party: the rust code added in order to build the TP in core do not have any external dependency but std.
* doing it externally: this was my original idea but, as stated also in the above post by @TheBlueMatt:
  *block template building is incredibly latency-sensitive, plus needs full access to the mempool and block template construction*

ccdle12 commented:
Just leaving a note that I'm currently working on an implementation of the template provider that integrates with the the rust sv2 project

UnidenifiedUser committed:
Update guix rust version to latest avaiable

ccdle12 commented:
We are currently working on changes, should be ready for review soon.

Fi3 commented:
we are implementing it in pure c++ without the use of the rust ffi lib so the PR is outdated
